File: dec24/gosolutions/day3/3.go
package day3

import (
	"bufio"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

var (
	p1 = regexp.MustCompile(`mul\(([0-9]?[0-9]?[0-9]),([0-9]?[0-9]?[0-9])\)`)
	p2 = regexp.MustCompile(`mul\(([0-9]?[0-9]?[0-9]),([0-9]?[0-9]?[0-9])\)|don't\(\).*?do\(\)|don't\(\)`)
)

func solveMul(input *string) int{
	matches := p1.FindAllStringSubmatch(*input, len(*input))
	res := 0
	for _, match := range matches {
		n1, _ := strconv.Atoi(match[1])
		n2, _ := strconv.Atoi(match[2])
		res += n1 * n2
	}

	return res
}

func solveMulConditional(input *string) int{
	matches := p2.FindAllStringSubmatch(*input, len(*input))
	res := 0
	for _, match := range matches {
		if strings.HasPrefix(match[0], "mul") {
			n1, _ := strconv.Atoi(match[1])
			n2, _ := strconv.Atoi(match[2])
			res += n1 * n2 
		} else if match[0] == "don't()" {
			break
		}
	}

	return res
}

func Day3() {
	// Get the path of the current Go source file
	_, currentFile, _, _ := runtime.Caller(0)

	// Get the directory of the Go source file
	currentDir := filepath.Dir(currentFile)
    parentDir := filepath.Dir(filepath.Dir(currentDir))
    file, err := os.Open(parentDir + "/testcases/3.txt")

    if err != nil {
        log.Fatal("Unable to open testcase file")
        return
    }

    defer file.Close()

    scanner := bufio.NewScanner(file)
	input := ""
	for scanner.Scan(){
		input += scanner.Text()
	}

	println("Expected answers:")
    println("Part a: ", 161)
    println("Part b: ", 48)
    println()
    println("Your answers:")
    println("Part a: ", solveMul(&input))
    println("Part b: ", solveMulConditional(&input))
}

File: cppsolutions/3.cpp
#include<iostream>
#include <fstream>
#include<vector>
#include <string>
#include <sstream>
#include <regex>
using namespace std;

// pattern match
// mul(X,Y)
long long solveMul(string &s){
    std::regex e("mul\\([0-9]?[0-9]?[0-9],[0-9]?[0-9]?[0-9]\\)");
    auto it = std::sregex_iterator(s.begin(), s.end(), e);
    auto end = std::sregex_iterator();
    long long res = 0;
    for(; it != end; ++it) {
        regex num("[0-9]?[0-9]?[0-9]");
        auto match = it->str();
        auto it2 = sregex_iterator(match.begin(), match.end(), num);
        auto end2 = sregex_iterator();
        long long currRes = 1;
        for(; it2 != end2; it2++){
            currRes *= stoi(it2->str());
        }
        
        res += currRes;
    }


    return res;
}

// pattern match
// mul(X,Y)
long long solveMulConditional(string &s){
    std::regex e("mul\\([0-9]?[0-9]?[0-9],[0-9]?[0-9]?[0-9]\\)|don't\\(\\).*?do\\(\\)|don't\\(\\)");
    auto it = std::sregex_iterator(s.begin(), s.end(), e);
    auto end = std::sregex_iterator();
    long long res = 0;
    for(; it != end; ++it) {
        auto match = it->str();
        // don't()...do()
        if(match.substr(0, 3) == "mul"){
            regex num("[0-9]?[0-9]?[0-9]");
            auto it2 = sregex_iterator(match.begin(), match.end(), num);
            auto end2 = sregex_iterator();
            long long currRes = 1;
            for(; it2 != end2; it2++){
                currRes *= stoi(it2->str());
            }
            res += currRes;
        } else if(match == "don't()"){
            break;
        }
    }

    return res;
}


int main() {
    ifstream f("../testcases/3.txt");

    if (!f) {
        cerr << "Error opening the file!" << endl;
        return 1;
    }

    string s;
    string input = "";
    while(getline(f, s)){
        input += s;
    }
    cout << "Part a: " << solveMul(input) << endl;
    cout << "Part b: " << solveMulConditional(input) << endl;

    // Expected answers:
    // Part a: 169021493
    // Part b: 111762583
    f.close();
    
    return 0;
}